### PR TITLE
Add `loop_labels` algorithm for TDT greedy decoding

### PR DESCRIPTION
**Pull request by artbataev**
# What does this PR do ?

TDT greedy decoding using`loop_labels` algorithm introduced in #7926.

**Collection**: [ASR]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py [hunk 1]
```diff
@@ -2819,7 +2823,7 @@ def forward(
 
         return (packed_result,)
 
-    def _greedy_decode_blank_as_pad(
+    def _greedy_decode_blank_as_pad_loop_frames(
```
Comment (hainan-xv) [inline nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:2826]: I am thinking maybe we could just delete this loop-frame version of decoding since it only works with TDT models trained with omega != 0.0; otherwise we will be having a lot of insertion errors.

**Review by hainan-xv — COMMENTED** [review]
great work! I left just a few small comments.

**Comment by artbataev** [comment]
Should be reviewed/merged after #8286

File: nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py [hunk 2]
```diff
@@ -2995,3 +2999,180 @@ def _greedy_decode_masked(
...
+
+        # loop while there are active indices
+        while (current_batch_size := active_indices.shape[0]) > 0:
+            # stage 1: get decoder (prediction network) output
+            decoder_output, state, *_ = self._pred_step(labels.unsqueeze(1), state, batch_size=current_batch_size)
+            decoder_output = self.joint.project_prednet(decoder_output)  # do not recalculate joint projection
+
+            # stage 2: get joint output, iteratively seeking for non-blank labels
+            # blank label in `labels` tensor means "end of hypothesis" (for this index)
+            time_indices_found = time_indices.clone().detach()  # preserve time indices for hyps, of batch_size size
```
[the remark this answers is not in the document]
Comment (artbataev) [reply nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py]: Here detach redundant, thanks. I fully reworked the implementation (no more this stuff :) )

File: nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+            time_indices_current_labels.copy_(time_indices, non_blocking=True)
+            if use_alignments:
+                if self.preserve_frame_confidence:
+                    logits = F.log_softmax(logits, dim=-1)
+                alignments.add_results_masked_(
+                    active_mask=active_mask,
+                    time_indices=time_indices_current_labels,
+                    logits=logits if self.preserve_alignments else None,
+                    labels=labels if self.preserve_alignments else None,
+                    confidence=self._get_confidence_tensor(logits) if self.preserve_frame_confidence else None,
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py:176]: @GNroy, is this correct for TDT model? 
Should it be `self._get_confidence_tensor(F.log_softmax(logits[:, :-duration_lengths]))`? or something else?

**Comment by artbataev** [comment]
jenkins

File: nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py (hunk 1 above)
Comment (hainan-xv) [reply nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py]: Yes that sounds good.

File: nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+            init_length=max_time * 2 if use_alignments else 1,  # blank for each timestep + text tokens
+            device=x.device,
+            float_dtype=x.dtype,
+            store_alignments=self.preserve_alignments,
+            store_frame_confidence=self.preserve_frame_confidence,
+        )
+
+        # durations
+        duration_lengths = len(self.durations)
+        all_durations = torch.tensor(self.durations, device=device).to(torch.long)
```
Comment (hainan-xv) [inline nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py:110]: Can we move this as a member of the class so that it doesn't need to be initialized every time?

**Review by hainan-xv — COMMENTED** [review]
Very minor comments.

File: nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py (hunk 3 above)
Comment (GNroy) [reply nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py:176]: Let's go with `self._get_confidence_tensor(F.log_softmax(logits[:, :-duration_lengths]))` for now.

File: nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py (hunk 4 above)
Comment (artbataev) [reply nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py]: I renamed  `duration_lengths` -> `num_durations`, and used tensor for `self.durations`. I agree there is no need to convert it to tensor here every time.
I think `num_durations` introduces no overhead for now, and I would prefer not to add denormalization to the class.
`self.durations` are stored on the CPU and moved to the other device only on demand. It is a safer version with mostly zero overhead (asynchronously moving several numbers to GPU is a lightweight operation compared to overall decoding). If we move the class member to GPU (e.g., `self.durations = self.durations.to(...)`), there will be undesirable side-effects in the multi-gpu environment (even if we convert this class to `nn.Module`, it is used through a "proxy" class that is not an `nn.Module`), e.g., `model.to("cuda:0"); model.transcribe(...); model.to("cuda:1")` – the tensor will be still on "cuda:0". I added a comment in the code.

File: nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py (hunk 3 above)
Comment (artbataev) [reply nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py]: Done, thanks!

**Comment by artbataev** [comment]
jenkins

File: nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py (hunk 4 above)
Comment (hainan-xv) [reply nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py]: Ah OK, this makes sense. Thanks for the clarification.

**Comment by artbataev** [comment]
jenkins

**Review by hainan-xv — APPROVED** [review]
Approved. This is a great algorithm for TDT decoding. Thanks!
